posts/go-cli-from-scratch.md

```diff
@@ -0,0 +1,768 @@
+```
+---
+authors:
+- Rodrigo Moran
+tags:
+- Golang
+- CLI
+date: 	2021-12-01T13:50:040Z
+title: "A No Framework Approach to Building A CLI"
+image:
+---
+```
+# No Framework Approach to Building A CLI with Go
+
+In this guide we will be building a CLI tool from scratch. No fancy frameworks or libraries instead we are building own highly minimal framework *loosely* based on [Cobra](https://cobra.dev/).
+
+Here's a taste of what we're building: **:fish: gupi**
+
+![Gupi Help Screen](../images/2022/03/help_screen.png "Gupi Help Screen")
+
+| :zap: Disclaimer: Popular CLI frameworks exist for every language, for most cases you're better off picking your favorite flavor and going with that. |
+| ------------------------------------------------------------ |
+
+Although if you've ever wondered how CLI Frameworks do what they do then read on. There are plenty of reasonable use cases for DIY tools, for a practical example look at Hashicorp (*Terraform, Nomad*). Most if not all their CLI products use [cli](https://github.com/mitchellh/cli) a Go CLI Framework developed by the founder.
+
+Cobra which I briefly name dropped above is argubably the best and most popular CLI Framework for Go. It is trusted by the Kubernetes and Docker community (**Kubectl** & **Docker CLI**) among many more notable projects: *Helm, Etcd, Istio...*
+
+Cobra has a deceptively simple API, for example here is a snippet directly from it's developer page:
+
+<img src="../images/2022/03/cobra_demo.png" alt="cobra_demo"  />
+
+This defines a single command *hello*, accepts a list of strings as arguments, and defines the logic to execute the command. The goal for this guide is to create a CLI with a similar look and feel.
+
+Here is the agenda for building **:fish: Gupi**:
+
+1. [Defining the requirements](#Defining-the-requirements)
+2. [Initializing the CLI](#Initializing-the-CLI)
+3. [Creating the Command struct](#Creating-the-Command-struct)
+4. [Wiring up components](#Wiring-up-components)
+5. [Adding subcommands](#Adding-subcommands): Add, Edit, List, Delete, Create, Version
+6. [Putting it all together](#Putting-it-all-together)
+
+Step 0: Make sure to have Go installed, see the official docs for the latest install instructions - [Go Install](https://go.dev/doc/install).
+
+You view the full source while following along at [Gupi](https://github.com/phantompunk/gupi).
+
+
+### Defining the requirements
+
+Keep in mind what we are building does not matter as much as *how* we are building it. It could easily be adapted to do whatever you need.
```

Comment:
```suggestion
Keep in mind that _what_ we are building does not matter as much as *_how_* we are building it. It could easily be adapted to do whatever you need.
```